src/libraries/System.Private.CoreLib/src/System/Threading/RegisteredWaitHandle.Portable.cs

```diff
@@ -2,17 +2,123 @@
+            if (resetEvent == null)
+            {
+                resetEvent = new AutoResetEvent(false);
+            }
+            return resetEvent;
+        }
+
+        private static void ReturnEvent(AutoResetEvent resetEvent)
+        {
+            if (Interlocked.CompareExchange(ref s_cachedEvent, resetEvent, null) != null)
+            {
+                resetEvent.Dispose();
+            }
+        }
+
+        private static readonly LowLevelLock s_callbackLock = new LowLevelLock();
+
+        /// <summary>
+        /// The callback to execute when the wait on <see cref="Handle"/> either times out or completes.
+        /// </summary>
+        internal _ThreadPoolWaitOrTimerCallback Callback { get; }
+
+        /// <summary>
+        /// The <see cref="SafeWaitHandle"/> that was registered.
+        /// </summary>
+        internal SafeWaitHandle Handle { get; }
+
+        /// <summary>
+        /// The time this handle times out at in ms.
+        /// </summary>
+        internal int TimeoutTimeMs { get; private set; }
+
+        internal int TimeoutDurationMs { get; }
+
+        internal bool IsInfiniteTimeout => TimeoutDurationMs == -1;
+
+        internal void RestartTimeout()
+        {
+            Debug.Assert(!IsInfiniteTimeout);
+            TimeoutTimeMs = Environment.TickCount + TimeoutDurationMs;
+        }
+
+        /// <summary>
+        /// Whether or not the wait is a repeating wait.
+        /// </summary>
+        internal bool Repeating { get; }
+
+        /// <summary>
+        /// The <see cref="WaitHandle"/> the user passed in via <see cref="Unregister(WaitHandle)"/>.
+        /// </summary>
+        private SafeWaitHandle? UserUnregisterWaitHandle { get; set; }
+
+        private IntPtr UserUnregisterWaitHandleValue { get; set; }
+
+        private static IntPtr InvalidHandleValue => new IntPtr(-1);
+
+        internal bool IsBlocking => UserUnregisterWaitHandleValue == InvalidHandleValue;
+
+        /// <summary>
+        /// The number of callbacks that are currently queued on the Thread Pool or executing.
+        /// </summary>
+        private int _numRequestedCallbacks;
+
+        /// <summary>
+        /// Notes if we need to signal the user's unregister event after all callbacks complete.
+        /// </summary>
+        private bool _signalAfterCallbacksComplete;
+
+        private bool _unregisterCalled;
+
+#pragma warning disable CS0414 // The field is assigned but its value is never used. Some runtimes may not support registered wait handles.
+        private bool _unregistered;
+#pragma warning restore CS0414
+
+        private AutoResetEvent? _callbacksComplete;
+
+        private AutoResetEvent? _removed;
+
         /// <summary>
         /// The <see cref="PortableThreadPool.WaitThread"/> this <see cref="RegisteredWaitHandle"/> was registered on.
         /// </summary>
         internal PortableThreadPool.WaitThread? WaitThread { get; set; }
 
+#if CORECLR
```

Comment:
This ifdef will go away once the unmanaged threadpool is deleted. Not worth to optimize factoring for it.